swarm/src/lib.rs

```diff
@@ -828,35 +893,17 @@ where TBehaviour: NetworkBehaviour,
             local_peer_id,
             transport,
             behaviour,
-            network: NetworkConfig::default(),
+            network_config: None,
         }
     }
 
-    pub fn incoming_limit(mut self, incoming_limit: usize) -> Self {
-        self.network.set_pending_incoming_limit(incoming_limit);
-        self
-    }
-
-    /// Sets the executor to use to spawn background tasks.
-    ///
-    /// By default, uses a threads pool.
-    pub fn executor(mut self, executor: impl Executor + Send + 'static) -> Self {
-        self.network.set_executor(Box::new(executor));
-        self
-    }
-
-    /// Shortcut for calling `executor` with an object that calls the given closure.
-    pub fn executor_fn(mut self, executor: impl Fn(Pin<Box<dyn Future<Output = ()> + Send>>) + Send + 'static) -> Self {
-        struct SpawnImpl<F>(F);
-        impl<F: Fn(Pin<Box<dyn Future<Output = ()> + Send>>)> Executor for SpawnImpl<F> {
-            fn exec(&self, f: Pin<Box<dyn Future<Output = ()> + Send>>) {
-                (self.0)(f)
-            }
-        }
-        self.network.set_executor(Box::new(SpawnImpl(executor)));
+    /// Configures the underlying network for the swarm.
+    pub fn network_config(mut self, cfg: NetworkConfig) -> Self {
```

Comment:
This doesn't look so great to me in terms of usability.

Instead of:
```rust
SwarmBuilder::new(...).executor(...).build()
```

One now has to do:
```rust
let mut cfg = NetworkConfig::default();
cfg.set_executor(...);
SwarmBuilder::new(...).network_config(cfg).build();
```

I've had that remark already for `PingConfig`.
It can really become very annoying to have to browse half of the documentation and import half of the library to create types that pretty much all the users of libp2p will have to create.

Reply:
Well, `libp2p-swarm` can certainly re-export things to make imports easier, but I really don't think essentially duplicating the `NetworkConfig` API on the `SwarmBuilder` is a better strategy, at least I wouldn't want to maintain that.

Reply:
I guess we could justify the API duplication by stating that, in principle, we want to keep open the possibility of `libp2p-swarm` to make predetermined (i.e. opinionated) choices on some of the underlying network configuration, even if we do not currently do so. A very similar justification could be made for not exposing the underlying `Network` of a `Swarm`, where additionally direct access to the network APIs may even cause potential state kept in a `Swarm` to become out-of-sync.

Reply:
> I guess we could justify the API duplication by stating that, in principle, we want to keep open the possibility of libp2p-swarm to make predetermined (i.e. opinionated) choices on some of the underlying network configuration, even if we do not currently do so.

That was indeed my reasoning.

The `Swarm` configuration and the `Network` configuration do intersect in some way, but aren't necessarily a subset or a superset of one another.

Reply:
Resolved by https://github.com/libp2p/rust-libp2p/pull/1519/commits/f90f548dfd81bb4da762e6bd870d947a3f9902e0.